activerecord/test/cases/finder_test.rb

```diff
@@ -693,6 +693,30 @@ def test_take_and_first_and_last_with_integer_should_return_an_array
     assert_kind_of Array, Topic.last(5)
   end
 
+  def test_first_and_last_with_limit_should_return_same_size_loaded_or_not
+    topics = Topic
+
+    assert_equal topics.limit(1).to_a.first(2).size, topics.limit(1).first(2).size
+    assert_equal topics.limit(2).to_a.first(2).size, topics.limit(2).first(2).size
+    assert_equal topics.limit(3).to_a.first(2).size, topics.limit(3).first(2).size
+
+    assert_equal topics.limit(1).to_a.last(2).size, topics.limit(1).last(2).size
+    assert_equal topics.limit(2).to_a.last(2).size, topics.limit(2).last(2).size
+    assert_equal topics.limit(3).to_a.last(2).size, topics.limit(3).last(2).size
+  end
+
+  def test_first_and_last_with_limit_should_return_same_size
+    topics = Topic
+
+    assert_equal topics.limit(1).first(2).size, topics.limit(1).last(2).size
+    assert_equal topics.limit(2).first(2).size, topics.limit(2).last(2).size
+    assert_equal topics.limit(3).first(2).size, topics.limit(3).last(2).size
```

Comment:
Previously `relation.to_a.first(n)`, `relation.to_a.last(n)`, and `relation.last(n)` works expected, but only `relation.first(n)` does not.
Consistency between `relation.to_a.last(n)` and `relation.last(n)` have already been covered by `test_last_on_relation_with_limit_and_offset`.
Whether `relation.first(n)` works correctly should compare `relation.to_a.first(n)` rather than `relation.last(n)`.